lib/floe/workflow/runner/kubernetes.rb

```diff
@@ -4,42 +4,102 @@ module Floe
   class Workflow
     class Runner
       class Kubernetes < Floe::Workflow::Runner
-        attr_reader :namespace, :server, :token
+        TOKEN_FILE   = "/run/secrets/kubernetes.io/serviceaccount/token"
+        CA_CERT_FILE = "/run/secrets/kubernetes.io/serviceaccount/ca.crt"
 
         def initialize(options = {})
           require "awesome_spawn"
           require "securerandom"
           require "base64"
+          require "kubeclient"
           require "yaml"
 
-          @namespace = options.fetch("namespace", "default")
-          @server    = options.fetch("server", nil)
-          @token     = options.fetch("token", nil)
-          @token   ||= File.read(options["token_file"]) if options.key?("token_file")
+          kubeconfig_file    = ENV.fetch("KUBECONFIG", nil) || options.fetch("kubeconfig", File.join(Dir.home, ".kube", "config"))
+          kubeconfig_context = options.fetch("kubeconfig_context", nil)
+
+          kubeconfig  = Kubeclient::Config.read(kubeconfig_file) if File.exist?(kubeconfig_file)
+          kubecontext = kubeconfig.context(kubeconfig_context) if kubeconfig
+
+          token   = options.fetch("token", nil)
+          token ||= File.read(options["token_file"]) if options.key?("token_file")
+          token ||= File.read(TOKEN_FILE) if File.exist?(TOKEN_FILE)
+
+          server   = options.fetch("server", nil)
+          server ||= URI::HTTPS.build(:host => ENV.fetch("KUBERNETES_SERVICE_HOST"), :port => ENV.fetch("KUBERNETES_SERVICE_PORT", 6443)) if ENV.key?("KUBERNETES_SERVICE_HOST")
+
+          ca_file   = options.fetch("ca_file", nil)
+          ca_file ||= CA_CERT_FILE if File.exist?(CA_CERT_FILE)
+
+          verify_ssl = options.fetch("verify_ssl", nil) == "false" ? OpenSSL::SSL::VERIFY_NONE : OpenSSL::SSL::VERIFY_PEER
+
+          if server && token
+            ssl_options = {:verify_ssl => verify_ssl}
+            ssl_options[:ca_file] = ca_file if ca_file
+
+            auth_options = {:bearer_token => token}
+          elsif kubecontext
+            server       = kubecontext.api_endpoint
+            ssl_options  = kubecontext.ssl_options
+            auth_options = kubecontext.auth_options
+          else
+            raise ArgumentError, "Missing connections options, provide a kubeconfig file or pass server and token via --docker-runner-options"
+          end
+
+          @namespace  = options.fetch("namespace", "default")
+
+          @kubeclient = Kubeclient::Client.new(server, "v1", :ssl_options => ssl_options, :auth_options => auth_options)
```

Comment:
Not necessarily for this PR, but...

This creates a new client every time - I wonder if it would be helpful to cache the client and reuse it if the parameters are the same?  I'm guessing that in most scenarios you're not sending these steps to multiple clusters.

Reply:
So the docker_runner is only set once, I was actually worried about the opposite problem where there is a connection issue and we have to re-open the client.  I was considering caching the server, ssl_options, and auth_options and being able to `self.kubeclient = nil` then reopen it on a connection issue.